### DISCUSSION TITLE
Timer ranges

### DISCUSSION THREAD

**Discussion by sciurius** [arendst/Tasmota#14525] [Ideas]
Timers can be used to switch a device on or off on preset times.

A common use for a timed device is to be active between two times. E.g., water the garden from 05:00 till 07:00.

For (mostly) lights, sunrise and sunset are relevant times. E.g., lights on from sunset to 23:00 and from 07:00 to sunrise.

You feel it coming: What if the sun rises before 07:00...? IIUC, this currently requires complex configuration using rules. I think this should be trivial since it is such a common case.

This could be done with configuring _ranges_ instead of _endpoints_. If the end point precedes the starting point, the range is ignored.

From UI perspective this could be very simple: just add a 'Range' checkbox to a timer and this will then consider this timer the starting poing, and the next timer the end point.

Does this make sense?

**Comment by rapi3** [comment]
It is very simple already to configure timers based on Sunset/Sunrise and required interval -/+ ( before or after ) from interface... check the documentation;
Sunrise/Sunset are calculated all the time automatically based of location and timezone configured by user.

![Screenshot_2022-01-20_14-13-47](https://user-images.githubusercontent.com/18531150/150345741-afe1f38f-a9e2-4c74-94f7-be68adc968de.png)
![Screenshot_2022-01-20_14-25-53](https://user-images.githubusercontent.com/18531150/150347429-bbca88cc-1a80-4568-a38d-dd4e57bc5e23.png)

**Reply by sfromis** [reply]
Where the complication comes is combining sunset/sunrise timing with fixed times taking control when the sunset/sunrise veers outside of this interval. This is manageable with a few rules, not very complex, but is more than merely having a sunset/sunrise timer.

**Comment by sciurius** [comment]
Exactly.

Moreover I think that the concept of ranges is more natural than isolated (independent!) on/off switching. You switch things on and later off again. Also, after a reboot, you can easily check if the current time is within some of the ranges, and act accordingly. You cannot do that with independent on/off switching.

**Reply by LucidEye** [reply]
Yes, this is a BIG fail-safe issue.  This is why durations, or time-spans are a more robust way of programming a timer... a "looping-timer" can check itself after a power failure or reboot and pick up where it left off on an action... a single-point timer can not do that.

**Comment by rapi3** [comment]
In my automatization I think I am using range at this moment with PulseTime.

**Comment by gemu2015** [comment]
i made a special script for timer ranges. only drawback is an overnight switch has to be split into 2 entries.
another more complex script supports visualization and any number of outputs
advantage of ranges in case of restart the appropriate state is immediately set

<img width="428" alt="Bildschirmfoto 2022-01-21 um 08 40 36" src="https://user-images.githubusercontent.com/11647075/150487205-74574958-88f4-42ca-b9ac-df906f51fbd5.png">

[compressed_script.txt](https://github.com/arendst/Tasmota/files/7911286/compressed_script.txt)
[color_script.pdf](https://github.com/arendst/Tasmota/files/7911287/color_script.pdf)
<img width="554" alt="Bildschirmfoto 2022-01-20 um 20 48 37" src="https://user-images.githubusercontent.com/11647075/150487426-a76f2413-5f62-41e0-97a4-90709947d182.png">

**Reply by ascillato2** [reply]
Nice !!!!

**Comment by sciurius** [comment]
Where can I get this?

**Reply by ascillato** [reply]
in the links above. between the 2 images.

**Reply by gemu2015** [reply]
the script is appended, you have to compile your own Tasmota version with scripting enabled.
you need the following additional defines.

#define USE_SCRIPT
#define USE_TIMERS
#define USE_SCRIPT_WEB_DISPLAY
#define SCRIPT_POWER_SECTION

**Reply by sciurius** [reply]
Ah, it's a custom version...

**Comment by Jason2866** [comment]
@sciurius  We provide special builds. There is a scripting variant for esp8266 / ESP32 which includes the above defines.
https://github.com/tasmota/install/tree/main/firmware/unofficial

**Reply by sciurius** [reply]
Interesting, thanks!

**Comment by sciurius** [comment]
@gemu2015 I understand that I need to define the ranges in the script itself?
Is it possible to define e.g. 15 minutes before sunrise?
What does (end)time `0r730` mean?

**Reply by gemu2015** [reply]
Currently no offset from sunset sunrise implemented, should be easy however will think about it soon
0r730 is a typo should read 0730

**Comment by sciurius** [comment]
Excuse me for asking... What is `{s}`, `{m}`, and `{e}` ?

**Reply by barbudor** [reply]
These are "macros" that are replaced at run time by HTML code:
```
{s} = <tr><th>
{m} = </th><td>
{e} = </td></tr>
```
It saves flash bytes

**Reply by sfromis** [reply]
They are used for column-formatting the sensors on the Tasmota front page, each line having tags for start,mid,end

**Comment by gemu2015** [comment]
ok i checked for adding offsets to sunrise, sunset.
it can be added easily but it increases script size that will no longer fit in the standard precompiled build.

with a custom build you can use option
#define USE_EEPROM
#define EEP_SCRIPT_SIZE 6200

this activates a special mode where you get 6200 bytes script size.
(standard script size is about 2560)

with a 4m Flash size you may also use the Filesystem
there you can set script size with e.g.
#define UFSYS_SIZE 4096

**Reply by sciurius** [reply]
Really? I just added/subtracted 15 minutes to/from sunset/rise ☺
![scrot20220122144137](https://user-images.githubusercontent.com/1118804/150640808-1b4c0460-af1f-4a3a-8c9e-5055c414a5e9.png)
````
if from==-61 {
from=sunrise-15
}
if from==-62 {
from=sunset+15
}
if to==-61 {
to=sunrise-15
}
if to==-62 {
to=sunset+15
}

````

**Reply by gemu2015** [reply]
That’s the easy way and may be ok
I made a version with individual offsets for every entry that is much more code

**Reply by sciurius** [reply]
Yes, that's more flexible... but why? There is no GUI, so you need to adjust the script anyway when you want to change timers.

**Reply by gemu2015** [reply]
what i did is doing some math in the descriptor line, so you can add offsets to sunrise or sunset
example:
=#ts("-1-5:-2+3:3:0")

so every single entry may be individually adjusted

i agree, mostly not necessary

**Reply by sciurius** [reply]
Are there real benefits of having a single parameter with colon-separated values over 4 values. E.g. `=#ts("1500:2100:2:5")` over `=#ts(1500,2100,2,5)` ?

**Reply by gemu2015** [reply]
both would be possible, however the separator is space not comma so 
=#ts(1500 2100 2 5)
would take 4 separate parameters 
may be this version could use less code space but would not allow this math =#ts("-1-5:-2+3:3:0")

in theory also web gui functions are available in scripter so input times etc are possible
eg this inputs a time
\>D
t1f="15:30"
tmin="0000"
tmax="2359"

\>W
tx(t1f "Timer 1 Start " "time" tmin tmax)

**Reply by gemu2015** [reply]
via a POST request.

**Reply by sciurius** [reply]
That must be one of the best kept secrets of Tasmota... No word in the docs.

**Comment by rapi3** [comment]
If you want to use rules it is not complicated:

Set your timer start time as you want and use to trigger rule with defined range.

Set your rule that will define pulsetime - interval for range and power on; something like this:
ON clock#Timer=1 DO backlog Pulsetime XX, Power1 ENDON
ON clock#Timer=2 DO backlog Pulsetime YY, Power1 ENDON
ON clock#Timer=3 DO backlog Pulsetime ZZ, Power1 ENDON
...
or if you have a device with UFS filesistem you can define/edit (Timer01....) and run from there the commands you want
ON clock#Timer=1 DO UfsRun Timer01 ENDON

if you also turn on/off manually or from another device then add:
ON Power1#state=0 DO Pulsetime 0 ENDON

**Comment by sciurius** [comment]
I notice that the scripts-enabled firmware does not send a STATE message when the state is changed by the script or manually. Observers will therefore have to wait until the periodic STATE message (once every 5 minutes) is sent.
The regulat firmware sends a STATE message immedeately.

Is this a feature or did I miss a magic setting?

FWIW: The Domoticz message **is** immedeately sent after a state change.

````
08:22:22.920 CMD: power 1
08:22:22.931 MQT: domoticz/in = {"idx":368,"nvalue":1,"svalue":"19","Battery":100,"RSSI":7}
08:22:22.934 MQT: stat/bulb02/RESULT = {"POWER":"ON"}
08:22:22.938 MQT: stat/bulb02/POWER = ON
08:26:38.373 MQT: tele/bulb02/STATE = {"Time":"2022-01-24T08:26:38","Uptime":"0T00:05:09","UptimeSec":309,"Heap":20,"SleepMode":"Dynamic","Sleep":10,"LoadAvg":42,"MqttCount":1,"POWER":"ON","Dimmer":19,"Color":"0631000000","HSBColor":"113,100,19","White":0,"CT":153,"Channel":[2,19,0,0,0],"Scheme":0,"Fade":"OFF","Speed":1,"LedTable":"ON","Wifi":{"AP":1,"SSId":"Oh Denneboom","BSSId":"xx:xx:xx:xx:xx:xx","Channel":11,"Mode":"11n","RSSI":84,"Signal":-58,"LinkCount":1,"Downtime":"0T00:00:03"}}
````

**Reply by barbudor** [reply]
The 3rd and 4th line of your logs are sent immediatly. It shows the state change as per the "stat" prefix included in the topic.
The 5th line is the periodic telemetry message ("tele" prefix) which is a recap sent only at periodic time
This is the standard behavior of Tasmota

I known nothing about the Domoticz integration but to be consistent with Tasmota behavior, the immediate message to Domoticz would be more logical that it includes only the changed states. So may be it is a requirement of Domoticz ?

**Comment by sciurius** [comment]
In this case it is HomeAssistant that doesn't pick up the state change until the STATE message is received.

I merely noticed the difference between the regular 10.1.0 firmware (that sends the STATE message immedeately) and the scripting enabled build 2022.01.3.

**Reply by gemu2015** [reply]
as @barbudor already said this is the normal behavior when switching power. as you see in the script we just invoke the tasmota power command here.
you may additionally publish a STATE cmd with the script.

btw just for fun i made a gui script for the timer setup.
this however needs a 4M device or the special build on 1M device mentioned above.

<img width="813" alt="Bildschirmfoto 2022-01-25 um 10 34 25" src="https://user-images.githubusercontent.com/11647075/150951392-2092c7a4-e045-4a38-9d2c-a8886596bef4.png">

**Reply by sfromis** [reply]
You should also be able to use [SetOption59](https://tasmota.github.io/docs/Commands/#setoption59) to get `.../STATE` published after commands like Power.

**Reply by i-perfectus** [reply]
can You share bin file?

**Reply by sciurius** [reply]
I try to understand the GUI...
The ±value doesn't make sense with fixed times, does it?
Would it be possible to combine the `time` and `hh:mm` fields into a single field that can contain either `sunrise`, `sunset`, or a hh:mm value? Or maybe something like `sr+0:15` (15 minutes after sunrise)?

**Reply by sfromis** [reply]
No, combining would lose a feature of the timers. As the docs about [Timers](https://tasmota.github.io/docs/Timers/) explain, the `+/-` window of up to 15 minutes allows for a random number to be added/subtracted. When you use Sunrise/Sunset, the hh/mm values below gets another `+/-` sign and becomes a fixed offset from the varying sun-dependent time, which you then can tweak with a bit of randomness too.

**Reply by gemu2015** [reply]
currently the offsets are valid for sunrise, sunset. if you select sunrise, sunset the appropriate times are filled into the time field. since the page is dynamic html under script control you could easily hide the offsets in case of fixed times.
the complete script is 3.6k which leaves about 2,5 k left for modifications.
since this implementation does not rely in any way to the Tasmota timers and also uses its own permanent storage you are completely free to design it as you like.

**Comment by gemu2015** [comment]
here is 1M max script size binary (in english) i included energy sensor for e.g. gosund
later i will provide some example scripts like this gosund one.

<img width="596" alt="Bildschirmfoto 2022-01-26 um 10 39 56" src="https://user-images.githubusercontent.com/11647075/151140020-baf17744-5603-4188-9d04-bb8fce9408fe.png">




[tasmota-1M_maxscript.bin.gz](https://github.com/arendst/Tasmota/files/7966229/tasmota-1M_maxscript.bin.gz)

**Comment by gemu2015** [comment]
current state of range timers:

<img width="725" alt="Bildschirmfoto 2022-01-30 um 15 33 08" src="https://user-images.githubusercontent.com/11647075/151706407-dd99c75e-a1c1-434a-a9ca-db4fbd5b60e2.png">
<img width="569" alt="Bildschirmfoto 2022-01-30 um 15 31 13" src="https://user-images.githubusercontent.com/11647075/151706416-129ea126-2b3e-44fb-b9ac-90c82a3d80f8.png">

**Reply by Jonsonmon** [reply]
> current state of range timers:
> 
> <img alt="Bildschirmfoto 2022-01-30 um 15 33 08" width="725" src="https://user-images.githubusercontent.com/11647075/151706407-dd99c75e-a1c1-434a-a9ca-db4fbd5b60e2.png">
> 
> <img alt="Bildschirmfoto 2022-01-30 um 15 31 13" width="569" src="https://user-images.githubusercontent.com/11647075/151706416-129ea126-2b3e-44fb-b9ac-90c82a3d80f8.png">

It looks fantastic and exactly what I need, could you please post the script you made with all the changes for sunrise offsets etc... ?

**Reply by gemu2015** [reply]
in this link you find a collection of special scripts like the above
https://www.dropbox.com/scl/fo/9ltbureyz71se5i7t4yvw/AJLcMsK98Cx_rhLQ3xA05eY?rlkey=uuxsn04d17lng6djkmtgie5kv&e=1&dl=0

**Comment by Swyter** [comment]
In case anyone else searches for this.

I made a tiny Berry script that reuses the Tasmota timer functionality, to turn a pair timers into a start-end range, shutting it down e.g. when a reboot happens and NTP time is still not configured (no Internet, wrong time/date on startup) and other fail-safe features that Tasmota lacks, like not shutting down something when there's a brownout and the end range timer never fires due to the device being inoperable at that second, even it it's just a small blip.

So it otherwise would stay on for the rest of the day, unlike even the most basic mechanical timer that would take care of this (which I find very silly and a regression of functionality for such a "smart" device).

So yeah, this adds that back:

> https://github.com/Swyter/tasmota32-berry-timer-interval

Doesn't currently support sunrise/sunset or random offsets, though, but it does what I need it to do.


---

To use it you just upload `autoexec.be` via the *Tools > Manage File system* web browser into the root folder. Then [edit it](https://github.com/Swyter/tasmota32-berry-timer-interval/blob/e1c9c82aafb908abcef7b5ff20d8562aba6065e2/autoexec.be#L164) to set which two timers are linked together and monitored.

By default it checks (and logs) if we're inside or outside the active interval every single minute, reacting accordingly. 
But you can change it [here](https://github.com/Swyter/tasmota32-berry-timer-interval/blob/e1c9c82aafb908abcef7b5ff20d8562aba6065e2/autoexec.be#L172).

For example, under *Configuration > Timers* I configure timer 1/`t1` as start of the range (on) and timer 2/`t2` as off. The first argument in the check function below is just to confirm that both should control the second/`2` Tasmota switch, it must match the "output" value in both timers.

`check_timer_interval_between(2, t1, t2)`

Not the prettiest, but seems to work well against real-life shenanigans. Hope others find it interesting.

**Reply by Jonsonmon** [reply]
> In case anyone else searches for this.
> 
> I made a tiny Berry script that reuses the Tasmota timer functionality, to turn a pair timers into a start-end range, shutting it down e.g. when a reboot happens and NTP time is still not configured (no Internet, wrong time/date on startup) and other fail-safe features that Tasmota lacks, like not shutting down something when there's a brownout and the end range timer never fires due to the device being inoperable at that second, even it it's just a small blip.
> 
> So it otherwise would stay on for the rest of the day, unlike even the most basic mechanical timer that would take care of this (which I find very silly and a regression of functionality for such a "smart" device).
> 
> So yeah, this adds that back:
> 
> > https://github.com/Swyter/tasmota32-berry-timer-interval
> 
> Doesn't currently support sunrise/sunset or random offsets, though, but it does what I need it to do.
> 
> To use it you just upload `autoexec.be` via the _Tools > Manage File system_ web browser into the root folder. Then [edit it](https://github.com/Swyter/tasmota32-berry-timer-interval/blob/e1c9c82aafb908abcef7b5ff20d8562aba6065e2/autoexec.be#L164) to set which two timers are linked together and monitored.
> 
> By default it checks (and logs) if we're inside or outside the active interval every single minute, reacting accordingly. But you can change it [here](https://github.com/Swyter/tasmota32-berry-timer-interval/blob/e1c9c82aafb908abcef7b5ff20d8562aba6065e2/autoexec.be#L172).
> 
> For example, under _Configuration > Timers_ I configure timer 1/`t1` as start of the range (on) and timer 2/`t2` as off. The first argument in the check function below is just to confirm that both should control the second/`2` Tasmota switch, it must match the "output" value in both timers.
> 
> `check_timer_interval_between(2, t1, t2)`
> 
> Not the prettiest, but seems to work well against real-life shenanigans. Hope others find it interesting.

Thanks for that
But I really need sunset/sunrise with offset...
